### PR TITLE
feat: Add EbeanLocalRelationshipWriterDAO and BaseLocalRelationshipBu…

### PR DESCRIPTION
**Pull request by zhixuanjia**
…ilder

Units are in next PR.

## Checklist

- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [x] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable)

### CODE REVIEW

File: dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanLocalRelationshipWriterDAO.java [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+          .setParameter(CommonColumnName.SOURCE, source.toString())
+          .setParameter(CommonColumnName.DESTINATION, destination.toString())
+          .setParameter(CommonColumnName.LAST_MODIFIED_ON, new Timestamp(now))
+          .setParameter(CommonColumnName.LAST_MODIFIED_BY, DEFAULT_ACTOR)
+          .execute();
+    }
+  }
+
+  @ParametersAreNonnullByDefault
+  private <RELATIONSHIP extends RecordTemplate> void processRemoveOption(String tableName, RELATIONSHIP relationship,
```
Comment (jsdonn) [inline dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanLocalRelationshipWriterDAO.java:121]: nit: processRemovalOption

File: dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanLocalRelationshipWriterDAO.java [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+      RemovalOption removalOption) {
+    SqlUpdate deletionSQL = _server.createSqlUpdate(SQLStatementUtils.deleteLocaRelationSQL(tableName, removalOption));
+    Urn source = getSourceUrnFromRelationship(relationship);
+    Urn destination = getDestinationUrnFromRelationship(relationship);
+
+    if (removalOption == RemovalOption.REMOVE_ALL_EDGES_FROM_SOURCE_TO_DESTINATION) {
```
Comment (jsdonn) [inline dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanLocalRelationshipWriterDAO.java:127]: switch/case?

**Review by jsdonn — CHANGES_REQUESTED** [review]
Mostly looks good, just some small things.

File: dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/utils/SQLStatementUtils.java [hunk 3]
```diff
@@ -26,6 +28,14 @@ public class SQLStatementUtils {
 
   private static final String SQL_URN_EXIST_TEMPLATE = "SELECT urn FROM %s WHERE urn = '%s'";
 
+  private static final String INSERT_LOCAL_RELATIONSHIP = "INSERT INTO %s (metadata, source, destination, source_type, "
+      + "destination_type, lastmodifiedon, lastmodifiedby) VALUE (:metadata, :source, :destination, :source_type,"
+      + " :destination_type, :lastmodifiedon, :lastmodifiedby)";
+
+  private static final String DELETE_BY_SOURCE = "DELETE FROM %s WHERE source = :source";
+  private static final String DELETE_BY_DESTINATION = "DELETE FROM %s WHERE destination = :destination";
+  private static final String DELETE_BY_SOURCE_AND_DESTINATION = "DELETE FROM %s WHERE destination = :destination ADN source = :source";
```
[the remark this answers is not in the document]
Comment (zhixuanjia) [reply dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/utils/SQLStatementUtils.java:37]: fixed.

File: dao-api/src/main/java/com/linkedin/metadata/dao/builder/BaseLocalRelationshipBuilder.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+/**
+ * Build a local relationship based on an ASPECT.
+ */
+public abstract class BaseLocalRelationshipBuilder<ASPECT extends RecordTemplate> {
+
+  private Class<ASPECT> _aspectClass;
+
+  @Value
+  public class LocalRelationUpdates {
```
[the remark this answers is not in the document]
Comment (zhixuanjia) [reply dao-api/src/main/java/com/linkedin/metadata/dao/builder/BaseLocalRelationshipBuilder.java:19]: good catch, updated.

File: dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/utils/SQLStatementUtils.java [hunk 5]
```diff
@@ -140,4 +150,27 @@ public static <ASPECT extends RecordTemplate> String createAspectBrowseSql(Strin
...
+   * @return SQL statement for inserting local relation.
+   */
+  public static String insertLocalRelationSQL(String tableName) {
+    return String.format(INSERT_LOCAL_RELATIONSHIP, tableName);
+  }
+
+  @Nonnull
+  @ParametersAreNonnullByDefault
+  public static String deleteLocaRelationSQL(final String tableName, final BaseGraphWriterDAO.RemovalOption removalOption) {
+    if (removalOption == BaseGraphWriterDAO.RemovalOption.REMOVE_ALL_EDGES_FROM_SOURCE) {
```
[the remark this answers is not in the document]
Comment (zhixuanjia) [reply dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/utils/SQLStatementUtils.java:166]: I don't see a big difference so leave it as it is.

File: dao-impl/ebean-dao/src/test/java/com/linkedin/metadata/dao/MysqlDevInstance.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
+package com.linkedin.metadata.dao;
```
[the remark this answers is not in the document]
Comment (zhixuanjia) [reply dao-impl/ebean-dao/src/test/java/com/linkedin/metadata/dao/MysqlDevInstance.java:1]: Sure, moved it to the utils folder.

File: dao-impl/ebean-dao/src/test/java/com/linkedin/metadata/dao/localrelationship/EbeanLocalRelationshipWriterDAOTest.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+        BarUrn.createFromString("urn:li:bar:789")));
+
+    List<BaseLocalRelationshipBuilder<AspectFooBar>.LocalRelationUpdates> updates = new BelongsToLocalRelationshipBuilder(AspectFooBar.class)
+        .buildRelationships(FooUrn.createFromString("urn:li:foo:123"), aspectFooBar);
+
+    _localRelationshipWriterDAO.processLocalRelationshipUpdates(updates);
+
+    List<SqlRow> rows = _server.createSqlQuery("select * from metadata_relationship_belongsto where destination='urn:li:foo:123'").findList();
+    assertEquals(rows.size(), 3);
+    _server.execute(Ebean.createSqlUpdate("delete from metadata_relationship_belongsto"));
```
[the remark this answers is not in the document]
Comment (zhixuanjia) [reply dao-impl/ebean-dao/src/test/java/com/linkedin/metadata/dao/localrelationship/EbeanLocalRelationshipWriterDAOTest.java:50]: yes, truncate works here too.

File: dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanLocalRelationshipWriterDAO.java (hunk 2 above)
Comment (zhixuanjia) [reply dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanLocalRelationshipWriterDAO.java:127]: I don't see a big difference so leave it as it is.

File: dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanLocalRelationshipWriterDAO.java (hunk 1 above)
Comment (zhixuanjia) [reply dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanLocalRelationshipWriterDAO.java:121]: updated

**Comment by jsdonn** [comment]
LGTM

**Review by jsdonn — APPROVED** [review]

**Pull request merged**